File: android/core/protocol/src/test/kotlin/fr/mina/gateway/protocol/ChatDeviceWrapTest.kt
package fr.mina.gateway.protocol

import org.json.JSONObject
import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertThrows
import org.junit.Test
import java.security.KeyFactory
import java.security.spec.PKCS8EncodedKeySpec
import java.security.spec.X509EncodedKeySpec

/**
 * Même vecteur que Node (`tests/fixtures/protocol/mina-chat-device-wrap-vectors.json`).
 * Si un côté change sa dérivation, ce test casse — au lieu d'un appairage qui échoue en
 * silence sur un vrai téléphone.
 */
class ChatDeviceWrapTest {
    private val vectors: JSONObject by lazy {
        val stream = javaClass.classLoader!!.getResourceAsStream("mina-chat-device-wrap-vectors.json")
            ?: error("vecteur d'enveloppement introuvable")
        JSONObject(stream.bufferedReader().readText())
    }

    private fun hex(value: String): ByteArray =
        value.chunked(2).map { it.toInt(16).toByte() }.toByteArray()

    private fun toHex(value: ByteArray): String =
        value.joinToString("") { "%02x".format(it) }

    private fun privateKey(hexValue: String) =
        KeyFactory.getInstance("EC").generatePrivate(PKCS8EncodedKeySpec(hex(hexValue)))

    private fun publicKey(hexValue: String) =
        KeyFactory.getInstance("EC").generatePublic(X509EncodedKeySpec(hex(hexValue)))

    @Test
    fun `derive exactement la meme cle que Node`() {
        val deviceId = vectors.getString("deviceId")
        val fromDevice = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
            deviceId = deviceId,
        )
        assertEquals(vectors.getString("expectedWrapKeyHex"), toHex(fromDevice))
    }

    @Test
    fun `les deux cotes convergent`() {
        val deviceId = vectors.getString("deviceId")
        val fromPc = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("pcPrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("devicePublicSpkiHex")),
            deviceId = deviceId,
        )
        val fromDevice = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
            deviceId = deviceId,
        )
        assertArrayEquals(fromPc, fromDevice)
    }

    @Test
    fun `un autre identifiant d appareil donne une autre cle`() {
        val samsung = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
            deviceId = "device-samsung",
        )
        val huawei = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
            deviceId = "device-huawei",
        )
        assertFalse(samsung.contentEquals(huawei))
    }

    @Test
    fun `la cle derivee ouvre reellement une cle d epoque enveloppee par le PC`() {
        val deviceId = vectors.getString("deviceId")
        val epochKey = ChatCrypto.randomKey()
        val pcSide = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("pcPrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("devicePublicSpkiHex")),
            deviceId = deviceId,
        )
        val wrapped = ChatCrypto.wrapEpochKey(pcSide, epochKey, deviceId, 1)

        val deviceSide = ChatCrypto.deriveDeviceWrapKey(
            privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
            peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
            deviceId = deviceId,
        )
        assertArrayEquals(epochKey, ChatCrypto.unwrapEpochKey(deviceSide, wrapped, deviceId, 1))
    }

    @Test
    fun `refuse une derivation sans identifiant`() {
        assertThrows(IllegalArgumentException::class.java) {
            ChatCrypto.deriveDeviceWrapKey(
                privateKey = privateKey(vectors.getString("devicePrivatePkcs8Hex")),
                peerPublicKey = publicKey(vectors.getString("pcPublicSpkiHex")),
                deviceId = "",
            )
        }
    }
}

File: tests/chat-device-wrap.test.mjs
import { createPrivateKey, createPublicKey, generateKeyPairSync, randomBytes } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { describe, expect, it } from 'vitest';
import { deriveDeviceWrapKey, unwrapEpochKey, wrapEpochKey } from '../src/devices/chat-crypto.mjs';

const vectors = JSON.parse(readFileSync(
  fileURLToPath(new URL('./fixtures/protocol/mina-chat-device-wrap-vectors.json', import.meta.url)),
  'utf8',
));

const priv = (hex) => createPrivateKey({ key: Buffer.from(hex, 'hex'), format: 'der', type: 'pkcs8' });
const pub = (hex) => createPublicKey({ key: Buffer.from(hex, 'hex'), format: 'der', type: 'spki' });

describe('clé d\'enveloppement dérivée par ECDH', () => {
  it('les deux côtés obtiennent la MÊME clé sans qu\'aucun secret ne transite', () => {
    const fromPc = deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: pub(vectors.devicePublicSpkiHex),
      deviceId: vectors.deviceId,
    });
    const fromDevice = deriveDeviceWrapKey({
      privateKey: priv(vectors.devicePrivatePkcs8Hex),
      peerPublicKey: pub(vectors.pcPublicSpkiHex),
      deviceId: vectors.deviceId,
    });
    expect(fromPc.toString('hex')).toBe(fromDevice.toString('hex'));
    // Vecteur figé : Kotlin lit le MÊME fichier, donc une divergence casserait ici aussi.
    expect(fromPc.toString('hex')).toBe(vectors.expectedWrapKeyHex);
  });

  it('un autre appareil obtient une clé différente — le sel est l\'identifiant', () => {
    const derive = (deviceId) => deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: pub(vectors.devicePublicSpkiHex),
      deviceId,
    }).toString('hex');
    expect(derive('device-huawei')).not.toBe(derive('device-samsung'));
  });

  it('une clé étrangère ne permet pas de retrouver la clé d\'enveloppement', () => {
    const intruder = generateKeyPairSync('ec', {
      namedCurve: 'P-256',
      privateKeyEncoding: { type: 'pkcs8', format: 'der' },
      publicKeyEncoding: { type: 'spki', format: 'der' },
    });
    const legitimate = deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: pub(vectors.devicePublicSpkiHex),
      deviceId: vectors.deviceId,
    });
    const forged = deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: createPublicKey({ key: intruder.publicKey, format: 'der', type: 'spki' }),
      deviceId: vectors.deviceId,
    });
    expect(forged.equals(legitimate)).toBe(false);
  });

  it('la clé dérivée transporte réellement une clé d\'époque de bout en bout', () => {
    const epochKey = randomBytes(32);
    const pcSide = deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: pub(vectors.devicePublicSpkiHex),
      deviceId: vectors.deviceId,
    });
    const wrapped = wrapEpochKey({
      deviceWrapKey: pcSide,
      epochKey: Buffer.from(epochKey),
      deviceId: vectors.deviceId,
      keyEpoch: 1,
    });

    const deviceSide = deriveDeviceWrapKey({
      privateKey: priv(vectors.devicePrivatePkcs8Hex),
      peerPublicKey: pub(vectors.pcPublicSpkiHex),
      deviceId: vectors.deviceId,
    });
    const opened = unwrapEpochKey({
      deviceWrapKey: deviceSide,
      wrapped,
      deviceId: vectors.deviceId,
      keyEpoch: 1,
    });
    expect(opened.equals(epochKey)).toBe(true);
  });

  it('refuse une dérivation sans identifiant d\'appareil', () => {
    expect(() => deriveDeviceWrapKey({
      privateKey: priv(vectors.pcPrivatePkcs8Hex),
      peerPublicKey: pub(vectors.devicePublicSpkiHex),
      deviceId: '',
    })).toThrow('device_wrap_device_id_manquant');
  });
});
